pyro/distributions/mixture.py

```diff
@@ -0,0 +1,136 @@
+    :param torch.constraints.Constraint constraint0: constraint that holds
+        wherever ``mask == 0``
+    :param torch.constraints.Constraint constraint1: constraint that holds
+        wherever ``mask == 1``
+    """
+    def __init__(self, mask, constraint0, constraint1):
+        self.mask = mask
+        self.constraint0 = constraint0
+        self.constraint1 = constraint1
+
+    def check(self, value):
+        result = self.constraint0.check(value)
+        mask = self.mask.expand(result.shape) if result.shape != self.mask.shape else self.mask
+        result[mask] = self.constraint1.check(value)[mask]
+        return result
+
+
+class MaskedMixture(TorchDistribution):
+    """
+    A masked deterministic mixture of two distributions.
+
+    This is useful when the mask is sampled from another distribution,
+    possibly correlated across the batch. Often the mask can be
+    marginalized out via enumeration.
+
+    Example::
+
+        change_point = pyro.sample("change_point",
+                                   dist.Categorical(torch.ones(len(data) + 1)),
+                                   infer={'enumerate': 'parallel'})
+        mask = torch.arange(len(data), dtype=torch.long) >= changepoint
+        with pyro.iarange("data", len(data)):
+            pyro.sample("obs", MaskedMixture(mask, dist1, dist2), obs=data)
+
+    :param torch.Tensor mask: A byte tensor toggling between ``component0``
+        and ``component1``.
+    :param pyro.distributions.TorchDistribution component0: a distribution
+    :param pyro.distributions.TorchDistribution component1: a distribution
+    """
+    arg_constraints = {}  # nothing can be constrained
+
+    def __init__(self, mask, component0, component1, validate_args=None):
+        if not torch.is_tensor(mask) or mask.dtype != torch.uint8:
+            raise ValueError('Expected mask to be a ByteTensor but got {}'.format(type(mask)))
+        if component0.event_shape != component1.event_shape:
+            raise ValueError('components event_shape disagree: {} vs {}'
+                             .format(component0.event_shape, component1.event_shape))
+        batch_shape = broadcast_shape(mask.shape, component0.batch_shape, component1.batch_shape)
+        if mask.shape != batch_shape:
+            mask = mask.expand(batch_shape)
+        if component0.batch_shape != batch_shape:
+            component0 = component0.expand(batch_shape)
+        if component1.batch_shape != batch_shape:
+            component1 = component1.expand(batch_shape)
+
+        self.mask = mask
+        self.component0 = component0
+        self.component1 = component1
+        super(MaskedMixture, self).__init__(batch_shape, component0.event_shape, validate_args)
+
+        # We need to disable _validate_sample on each component since samples are only valid on the
+        # component from which they are drawn. Instead we perform validation using a MaskedConstraint.
```

Comment:
neat

Reply:
It seems kludgey to me, but I haven't seen a way around it. Let's keep this shortcoming in mind as we evolve the `torch.distributions` validation patterns. cc @alicanb @vishwakftw